src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -0,0 +1,409 @@
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", "lfuda", "minAvgWeight:cache", cacheLocation);
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return {};
+  } catch(std::exception &e) {
+    return -1;
+  }
+  
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", "lfuda", "minAvgWeight:weight", boost::lexical_cast<int>(weight));
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return {};
+
+    return std::get<0>(resp).value(); /* Returns number of fields set */
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+int LFUDAPolicy::get_min_avg_weight(optional_yield y) {
+  response<int> resp;
+
+  try {
+    boost::system::error_code ec;
+    request req;
+    req.push("HEXISTS", "lfuda", "minAvgWeight:cache");
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec)
+      return -1;
+  } catch(std::exception &e) {
+    return -1;
+  }
+
+  if (!std::get<0>(resp).value()) {
+    // Is int_max what we want here? -Sam
+    if (set_min_avg_weight(0, dir->cct->_conf->rgw_local_cache_address, y)) { /* Initialize minimum average weight */
+      return 0;
+    } else {
+      return -1;
+    }
+  }
+
+  try { 
+    boost::system::error_code ec;
+    response<std::string> value;
+    request req;
+    req.push("HGET", "lfuda", "minAvgWeight:weight");
+      
+    redis_exec(conn, ec, req, value, y);
+
+    if (ec)
+      return -1;
+
+    return std::stoi(std::get<0>(value).value());
+  } catch(std::exception &e) {
+    return -1;
+  }
+}
+
+CacheBlock LFUDAPolicy::find_victim(const DoutPrefixProvider* dpp, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
+  if (entries_map.empty())
+    return {};
+
+  auto it = std::min_element(std::begin(entries_map), std::end(entries_map),
+			      [](const auto& l, const auto& r) { return l.second->localWeight < r.second->localWeight; });
+
+  /* Get victim cache block */
+  CacheBlock victim;
+  victim.cacheObj.objName = it->second->key;
+  victim.cacheObj.bucketName = cacheNode->get_attr(dpp, victim.cacheObj.objName, "bucket_name", y); // generalize for other cache backends -Sam
+  victim.blockID = 0; // find way to get ID -Sam 
+
+  if (dir->get(&victim, y) < 0) {
+    return {};
+  }
+
+  return victim;
+}
+
+void LFUDAPolicy::shutdown() {
+  dir->shutdown();
+  
+  // call cancel() on the connection's executor
+  boost::asio::dispatch(conn->get_executor(), [c = conn] { c->cancel(); });
+}
+
+int LFUDAPolicy::exist_key(std::string key, optional_yield y) {
+  if (entries_map.count(key) != 0) {
+    return true;
+  }
+
+  return false;
+}
+
+int LFUDAPolicy::get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheNode, optional_yield y) {
+  response<std::string> resp;
+  int age = get_age(y);
+
+  if (exist_key(block->cacheObj.objName, y)) { /* Local copy */
```

Comment:
We need to search for the block and not the object in case of LFUDA as we are storing blocks in case of a read cache.

Reply:
It should be blockId

Reply:
I changed it to be the same key as what the directory holds, which uses the blockID. However, the representation in the cache node itself uses just the object name. Should this also be changed?

Reply:
Where exactly is this? Can you point out the line number?

Reply:
For example, line 221.

Reply:
I believe I should add an `oid_in_cache` value to the cache_block structure to achieve this because otherwise, I'm not sure how I could use the `oid_in_cache` as formed in the filter driver without passing it somehow to the method.

Reply:
`oid_in_cache` is formed by concatenating the object name, bucket name, block offset and length. I believe `block` has all these variables set. You can form the `key` here using the same logic